### ISSUE TITLE
failed to load tokenizer from file 

### ISSUE THREAD

**Issue by unrea1-sama** [huggingface/tokenizers#862]
I'm now training my own tokenizer. When I set ```continuing_subword_prefix``` and train the tokenizer, the tokenizer can't load from the saved file. The code is listed below:
```python
from tokenizers import Tokenizer
from tokenizers.models import BPE
from tokenizers.trainers import BpeTrainer
from tokenizers.pre_tokenizers import Whitespace, Digits, Sequence

PAD = "[PAD]"
BOS = "[BOS]"
EOS = "[EOS]"
UNK = "[UNK]"
special_tokens = [PAD, BOS, EOS, UNK]
tokenizer = Tokenizer(BPE(unk_token=UNK,continuing_subword_prefix='##'))
trainer = BpeTrainer(special_tokens=special_tokens)
tokenizer.pre_tokenizer = Sequence([Whitespace(), Digits(individual_digits=True)])
tokenizer.train(files=["corpus/en"],trainer=trainer)

tokenizer.save("corpus/tokenizer.json")

tokenizer.from_file("corpus/tokenizer.json")
```

when executing ```tokenizer.from_file("corpus/tokenizer.json")```, and error occured:
```bash
thread '<unnamed>' panicked at 'byte index 2 is out of bounds of `h`', /__w/tokenizers/tokenizers/tokenizers/src/models/bpe/model.rs:155:53
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Traceback (most recent call last):
  File "learn_tokenizer.py", line 18, in <module>
    tokenizer.from_file("corpus/tokenizer.json")
pyo3_runtime.PanicException: byte index 2 is out of bounds of `h`
```
I wonder if it's a bug or my misuse. Thankyou!

**Comment by Narsil** [comment]
This shouldn't be linked to `continuing_subword_prefix`. Is the JSON readable from python itslef `with open(filename, 'r') as f: json.load(f)` ?

If not it means the JSON is corrupted. Could be you are out of space or some very odd bug in the library.

If the JSON is readable and doesn't work in `tokenizers`, do you mind sharing it so we can take a look ?

**Comment by unrea1-sama** [comment]
> This shouldn't be linked to `continuing_subword_prefix`. Is the JSON readable from python itslef `with open(filename, 'r') as f: json.load(f)` ?
> 
> If not it means the JSON is corrupted. Could be you are out of space or some very odd bug in the library.
> 
> If the JSON is readable and doesn't work in `tokenizers`, do you mind sharing it so we can take a look ?

The json is readable from python, however the tokenizer crashed when loading the tokenizer from the file. I put the json file in the attachment ([tokenizer.txt](https://github.com/huggingface/tokenizers/files/7792275/tokenizer.txt), since github doesn't support json, I changed the extension to txt). 

I also find that without a suffix, the BPE tokenizer can't decode the encoded inputs properly. So when training the tokenizer, I have to explicitly set 
```python
trainer = BpeTrainer(special_tokens=special_tokens,end_of_word_suffix="</w>")
```
and use a BPE decoder, like:
```python
from tokenizers import decoders
tokenizer.decoder = decoders.BPEDecoder()

``` 
However, I didn't find any instruction about how to use the decoder provided in the library (or I just missed it in the document), which makes training a new tokenizer and properly use it difficult. And the class `BPE`, `BpeTrainer` both have parameter named `continuing_subword_prefix` and `end_of_word_suffix`, which is confusing. If I set both of them with different string, will they conflict?

**Comment by Narsil** [comment]
Yes that's exactly it:

`continuing_subword_prefix` is defined both in the model and in the trainer. However there seems to be an error in the flow of information.

Ideally we would get the one you define, regardless if it's in the trainer or the model, but right now it seems we can trigger inconsistent state with your code, where `BPE` has `##` as the continuing subword prefix, where as the trainer did not and so the merges saved in the file are inconsistent with `continuing_subword_prefix`.

The bug is that we try to suppress the prefix when joining the merges in order to build the internal MergeMap. but the prefix is inexistent for these.

tl;dr
```python
from tokenizers import Tokenizer
from tokenizers.models import BPE
from tokenizers.trainers import BpeTrainer
from tokenizers.pre_tokenizers import Whitespace, Digits, Sequence

PAD = "[PAD]"
BOS = "[BOS]"
EOS = "[EOS]"
UNK = "[UNK]"
special_tokens = [PAD, BOS, EOS, UNK]
tokenizer = Tokenizer(BPE(unk_token=UNK))
trainer = BpeTrainer(special_tokens=special_tokens,continuing_subword_prefix='##')
tokenizer.pre_tokenizer = Sequence([Whitespace(), Digits(individual_digits=True)])
tokenizer.train(files=["corpus/en"],trainer=trainer)

tokenizer.save("corpus/tokenizer.json")

tokenizer.from_file("corpus/tokenizer.json")
```
 should work.

Looking into proper fixes though.

**Comment by unrea1-sama** [comment]
Thankyou!
